File: app/src/main/java/com/example/administrator/androidtest/widget/pdfview/PdfRenderView.java
package com.example.administrator.androidtest.widget.pdfview;

import android.annotation.SuppressLint;
import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.Color;
import android.graphics.drawable.GradientDrawable;
import android.graphics.pdf.PdfRenderer;
import android.os.Handler;
import android.os.Looper;
import android.os.ParcelFileDescriptor;
import android.util.AttributeSet;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.recyclerview.widget.DividerItemDecoration;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.example.liblog.SLog;

import java.io.File;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.Future;

public class PdfRenderView extends RecyclerView {
    private static final String TAG = "PdfRenderView";

    private int dividerHeight = 10;
    private String currentPath = "";
    private PdfRenderer pdfRenderer;
    private final PdfBitmapPool pdfBitmapPool = new PdfBitmapPool();
    private final PdfBitmapCache pdfBitmapCache = new PdfBitmapCache();
    private final Map<String, Bitmap> activeBitmapCache = new ConcurrentHashMap<>();
    private ExecutorService executor = Executors.newCachedThreadPool();
    private final Handler mainHandler = new Handler(Looper.getMainLooper());

    public PdfRenderView(Context context) {
        this(context, null);
    }

    public PdfRenderView(Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
        pdfBitmapCache.setRecyclerCallback(pdfBitmapPool::put);
        initView();
    }

    private void initView() {
        setLayoutManager(new LinearLayoutManager(getContext(), LinearLayoutManager.VERTICAL, false));
        initDecoration(dividerHeight);
    }

    private void initDecoration(int height) {
        for (int index = getItemDecorationCount() - 1; index > 0; index--) {
            removeItemDecorationAt(index);
        }
        DividerItemDecoration decoration = new DividerItemDecoration(getContext(), DividerItemDecoration.VERTICAL);
        GradientDrawable drawable = new GradientDrawable();
        drawable.setSize(0, height);
        drawable.setColor(Color.TRANSPARENT);
        drawable.setShape(GradientDrawable.RECTANGLE);
        decoration.setDrawable(drawable);
        addItemDecoration(decoration);
    }

    public void load(String path) {
        try {
            if (currentPath.equals(path)) {
                SLog.d(TAG, "load: use same path. path = " + path);
                return;
            }
            currentPath = path;
            if (getAdapter() != null && getAdapter() instanceof PdfItemAdapter) {
                ((PdfItemAdapter) getAdapter()).cancelJob();
            }
            closePdf();
            activeBitmapCache.clear();
            File file = new File(path);
            int startIndex = path.lastIndexOf("/") + 1;
            int endIndex = path.lastIndexOf(".");
            String prefixKey = path.substring(startIndex, endIndex);
            prefixKey = prefixKey.isEmpty() ? "pdf-" + file.lastModified() : prefixKey + "-" + file.lastModified();
            ParcelFileDescriptor pfd = ParcelFileDescriptor.open(new File(path), ParcelFileDescriptor.MODE_READ_ONLY);
            pdfRenderer = new PdfRenderer(pfd);
            SLog.d(TAG, "load: path = " + path + ", prefixKey = " + prefixKey + ", lastModified = " + file.lastModified() + ", pageCount = " + pdfRenderer.getPageCount());
            setAdapter(new PdfItemAdapter(prefixKey));
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    @Override
    protected void onAttachedToWindow() {
        super.onAttachedToWindow();
        pdfBitmapCache.setRecyclerCallback(pdfBitmapPool::put);
        if (executor.isShutdown()) {
            executor = Executors.newCachedThreadPool();
        }
    }

    @Override
    protected void onDetachedFromWindow() {
        super.onDetachedFromWindow();
        SLog.d(TAG, "onDetachedFromWindow");
        closePdf();
        activeBitmapCache.clear();
        pdfBitmapCache.clear();
        pdfBitmapPool.clear();
        executor.shutdown();
        currentPath = "";
    }

    private void closePdf() {
        try {
            synchronized (PdfRenderer.class) {
                if (pdfRenderer != null) {
                    pdfRenderer.close();
                    pdfRenderer = null;
                }
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    @SuppressLint("NotifyDataSetChanged")
    public void setDividerHeight(int height) {
        dividerHeight = height;
        initDecoration(height);
        if (getAdapter() != null) {
            getAdapter().notifyDataSetChanged();
        }
    }

    private class PdfItemAdapter extends RecyclerView.Adapter<PdfItemViewHolder> {

        private int pdfPageCount;
        private final String prefixKey;

        public PdfItemAdapter(String key) {
            prefixKey = key;
            if (pdfRenderer != null) {
                pdfPageCount = pdfRenderer.getPageCount();
            }
        }

        @NonNull
        @Override
        public PdfItemViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
            ImageView iv = new ImageView(parent.getContext());
            iv.setLayoutParams(new ViewGroup.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT));
            iv.setScaleType(ImageView.ScaleType.CENTER_INSIDE);
            return new PdfItemViewHolder(iv);
        }

        @Override
        public void onBindViewHolder(@NonNull PdfItemViewHolder holder, int position) {
            holder.bindPdfItem(position, prefixKey + position);
        }

        @Override
        public void onViewRecycled(@NonNull PdfItemViewHolder holder) {
            super.onViewRecycled(holder);
            String cacheKey = prefixKey + holder.getAbsoluteAdapterPosition();
            Bitmap bitmap = activeBitmapCache.remove(cacheKey);
            SLog.d(TAG, "onViewRecycled: cacheKey = " + cacheKey + ", bitmap = " + bitmap);
            if (bitmap != null) {
                pdfBitmapCache.put(cacheKey, bitmap);
            }
        }

        @Override
        public int getItemCount() {
            return pdfPageCount;
        }

        public void cancelJob() {
            int childCount = getChildCount();
            for (int i = 0; i < childCount; i++) {
                ViewHolder viewHolder = getChildViewHolder(getChildAt(i));
                if (viewHolder instanceof PdfItemViewHolder) {
                    ((PdfItemViewHolder)viewHolder).cancelJob();
                }
            }
        }
    }

    private class PdfItemViewHolder extends RecyclerView.ViewHolder {

        private final ImageView imageView;
        private Future<?> futureJob;

        public PdfItemViewHolder(@NonNull View itemView) {
            super(itemView);
            imageView = (ImageView) itemView;
        }

        public void bindPdfItem(int position, String cacheKey) {
            imageView.setTag(position);
            Bitmap cacheBitmap = activeBitmapCache.get(cacheKey);
            if (cacheBitmap != null && !cacheBitmap.isRecycled()) {
                SLog.d(TAG, "bindPdfItem: use active cache: position = " + position + ", cacheKey = " + cacheKey);
                imageView.setImageBitmap(cacheBitmap);
                return;
            }
            cacheBitmap = pdfBitmapCache.remove(cacheKey);
            if (cacheBitmap != null && !cacheBitmap.isRecycled()) {
                SLog.d(TAG, "bindPdfItem: use lru cache: position = " + position + ", cacheKey = " + cacheKey);
                activeBitmapCache.put(cacheKey, cacheBitmap);
                imageView.setImageBitmap(cacheBitmap);
                return;
            }
            if (pdfRenderer != null) {
                SLog.d(TAG, "bindPdfItem: use source: position = " + position + ", cacheKey = " + cacheKey);
                loadPdfBitmap(position, bitmap -> {
                    SLog.d(TAG, "bindPdfItem: load bitmap success: position = " + position + ", tag = " + imageView.getTag());
                    if (imageView.getTag().equals(position)) {
                        activeBitmapCache.put(cacheKey, bitmap);
                        mainHandler.post(() -> imageView.setImageBitmap(bitmap));
                    } else {
                        pdfBitmapCache.put(cacheKey, bitmap);
                    }
                });
            }
        }

        public void cancelJob() {
            if (futureJob != null) {
                futureJob.cancel(true);
            }
        }

        private void loadPdfBitmap(int position, LoadPdfBitmapCallback callback) {
            futureJob = executor.submit(() -> {
                if (imageView.getTag().equals(position)) {
                    // java.lang.IllegalStateException: Current page not closed
                    synchronized (PdfRenderer.class) {
                        if (pdfRenderer != null) {
                            PdfRenderer.Page page = pdfRenderer.openPage(position);
                            int width = getResources().getDisplayMetrics().densityDpi * page.getWidth() / 72;
                            int height = getResources().getDisplayMetrics().densityDpi * page.getHeight() / 72;
                            int pageWidth = page.getWidth();
                            int pageHeight = page.getHeight();
                            int viewWidth = getWidth();
                            if (pageWidth > 0) {
                                width = viewWidth;
                                height = (int) (viewWidth * 1f / pageWidth * pageHeight);
                            }
                            Bitmap bitmap = pdfBitmapPool.get(width, height);
                            page.render(bitmap, null, null, PdfRenderer.Page.RENDER_MODE_FOR_DISPLAY);
                            page.close();
                            SLog.d(TAG, "loadPdfBitmap: position = " + position + ", page.getWidth() = " + page.getWidth() + ", page.getHeight() = " + page.getHeight() + ", width = " + width + ", height = " + height + ", densityDpi = " + getResources().getDisplayMetrics().densityDpi);
                            if (callback != null) {
                                callback.onLoadPdfBitmap(bitmap);
                            }
                        }
                    }
                }
            });
        }
    }

    private interface LoadPdfBitmapCallback {
        void onLoadPdfBitmap(Bitmap bitmap);
    }
}

File: app/src/main/java/com/example/administrator/androidtest/widget/pdfview/PdfRenderView.kt
package com.example.administrator.androidtest.widget.pdfview

import android.annotation.SuppressLint
import android.content.Context
import android.graphics.Bitmap
import android.net.Uri
import android.util.AttributeSet
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.example.administrator.androidtest.widget.pdfview.pdfapi.PdfApi
import com.example.administrator.androidtest.widget.pdfview.pdfapi.PdfApiFactory
import com.example.administrator.androidtest.widget.pdfview.pdfapi.PdfApiType
import com.example.administrator.androidtest.widget.pdfview.pdfapi.addDivider
import com.example.administrator.androidtest.widget.pdfview.pdfapi.isValid
import com.example.administrator.androidtest.widget.pdfview.pdfapi.md5
import com.example.liblog.SLog
import kotlinx.coroutines.CoroutineExceptionHandler
import kotlinx.coroutines.CoroutineName
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import java.io.File

class PdfRenderView @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyleAttr: Int = 0
) : RecyclerView(context, attrs, defStyleAttr) {

    private var dividerHeight = 10
    private var currentTag = ""
    private val pdfApi = PdfApiFactory.create(context, PdfApiType.ANDROID)
    private val jobSet = hashSetOf<Job>()
    private var coroutineScope = getPdfRenderCoroutineScope()

    init {
        initView()
    }

    private fun initView() {
        layoutManager = LinearLayoutManager(context, LinearLayoutManager.VERTICAL, false)
        addDivider(dividerHeight)
    }

    suspend fun load(path: String, tag: String = "") {
        try {
            currentTag = tag.ifEmpty {
                val md5 = File(path).md5()
                if (md5.isNotEmpty() && currentTag == md5) {
                    SLog.d(TAG, "load: skipping unchanged file, md5 = $md5")
                    return
                } else {
                    md5
                }
            }
            cancelJob()
            closePdf()
            pdfApi.load(File(path), currentTag)
            adapter = PdfItemAdapter()
            SLog.d(TAG, "load: PDF loaded from path successfully, currentTag = $currentTag, pageCount = ${pdfApi.getPageCount()}")
        } catch (e: Exception) {
            SLog.e(TAG, "load: exception message: ${e.message}")
        }
    }

    suspend fun load(uri: Uri, tag: String = "") {
        try {
            currentTag = tag.ifEmpty {
                val md5 = uri.md5(context)
                if (md5.isNotEmpty() && currentTag == md5) {
                    SLog.d(TAG, "load: skipping unchanged file, md5 = $md5")
                    return
                } else {
                    md5
                }
            }
            cancelJob()
            closePdf()
            pdfApi.load(uri, currentTag)
            adapter = PdfItemAdapter()
            SLog.d(TAG, "load: PDF loaded from uri successfully, currentTag = $currentTag, pageCount = ${pdfApi.getPageCount()}")
        } catch (e: Exception) {
            SLog.e(TAG, "load: exception message: ${e.message}")
        }
    }

    override fun onAttachedToWindow() {
        super.onAttachedToWindow()
        if (!coroutineScope.isActive) {
            coroutineScope = getPdfRenderCoroutineScope()
        }
    }

    override fun onDetachedFromWindow() {
        SLog.d(TAG, "onDetachedFromWindow： enter")
        super.onDetachedFromWindow()
        cancelJob()
        closePdf()
    }

    private fun cancelJob() {
        jobSet.forEach {
            it.cancel()
        }
        jobSet.clear()
    }

    private fun closePdf() {
        synchronized(PdfApi::class.java) {
            try {
                pdfApi.close()
            } catch (e: Exception) {
                SLog.e(TAG, "closePdf: exception message: ${e.message}")
            }
        }
    }

    @SuppressLint("NotifyDataSetChanged")
    fun setDividerHeight(height: Int) {
        dividerHeight = height
        addDivider(height)
        adapter?.notifyDataSetChanged()
    }

    private fun getPdfRenderCoroutineScope() = CoroutineScope(SupervisorJob() + Dispatchers.Main.immediate + CoroutineExceptionHandler { _, exception ->
        SLog.e(TAG, "catch PdfRenderView coroutine exception: ${exception.message}")
    })

    private inner class PdfItemAdapter : Adapter<PdfItemViewHolder>() {
        private var pdfPageCount = pdfApi.getPageCount()

        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): PdfItemViewHolder {
            val iv = ImageView(parent.context)
            iv.layoutParams = ViewGroup.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT)
            iv.scaleType = ImageView.ScaleType.CENTER_INSIDE
            return PdfItemViewHolder(iv)
        }

        override fun onBindViewHolder(holder: PdfItemViewHolder, position: Int) {
            holder.bindPdfItem(position)
        }

        override fun onViewRecycled(holder: PdfItemViewHolder) {
            super.onViewRecycled(holder)
            val bitmap = pdfApi.removeUsedBitmap(holder.layoutPosition)
            SLog.d(TAG, "onViewRecycled: layoutPosition = ${holder.layoutPosition}, bitmap = $bitmap")
            if (bitmap.isValid()) {
                pdfApi.putCachedBitmap(holder.layoutPosition, pdfApi.getBitmapSize(), bitmap)
            }
        }

        override fun getItemCount() = pdfPageCount
    }

    private inner class PdfItemViewHolder(itemView: View) : ViewHolder(itemView) {
        private val imageView = itemView as ImageView

        fun bindPdfItem(position: Int) {
            imageView.tag = position
            val cacheBitmap = pdfApi.getMemoryBitmap(position)
            if (cacheBitmap.isValid()) {
                SLog.d(TAG, "bindPdfItem: use memory cache: position = $position")
                imageView.setImageBitmap(cacheBitmap)
                return
            }
            SLog.d(TAG, "bindPdfItem: load bitmap start: position = $position")
            val job = coroutineScope.launch(CoroutineName("Job-$position")) {
                val bitmap = loadPdfBitmap(position)
                SLog.d(TAG, "bindPdfItem: load bitmap success: position = $position, tag = ${imageView.tag}, isValid = ${bitmap.isValid()}")
                if (bitmap.isValid()) {
                    if (imageView.tag == position) {
                        pdfApi.putUsedBitmap(position, pdfApi.getBitmapSize(), bitmap)
                        imageView.setImageBitmap(bitmap)
                    } else {
                        pdfApi.putCachedBitmap(position, pdfApi.getBitmapSize(), bitmap)
                    }
                }
            }
            job.invokeOnCompletion {
                if (it != null) {
                    SLog.d(TAG, "invokeOnCompletion: exception message: ${it.message}")
                }
                jobSet.remove(job)
            }
            jobSet.add(job)
        }

        private suspend fun loadPdfBitmap(position: Int): Bitmap? = withContext(Dispatchers.IO) {
            if (imageView.tag == position) {
                // java.lang.IllegalStateException: Current page not closed
                synchronized(PdfApi::class.java) {
                    val bitmap = pdfApi.getDiskOrNewBitmap(position)
                    return@withContext bitmap
                }
            }
            return@withContext null
        }
    }

    companion object {
        private const val TAG = "PdfRenderView"
    }
}
